h/services/flag.py

```diff
@@ -28,6 +28,25 @@ def flagged(self, user, annotation):
         query = self.session.query(models.Flag).filter_by(user=user, annotation=annotation)
         return query.count() > 0
 
+    def all_flagged(self, user, annotation_ids):
+        """
+        Check which of the given annotation IDs the given user has flagged.
+
+        :param user: The user to check for a flag.
+        :type user: h.models.User
+
+        :param annotation_ids: The IDs of the annotations to check.
+        :type annotation_ids: sequence of unicode
+
+        :returns The subset of the IDs that the given user has flagged.
+        :rtype set of unicode
+        """
+        query = self.session.query(models.Flag.annotation_id) \
+                            .filter(models.Flag.annotation_id.in_(annotation_ids),
```

Comment:
A very small thing, but it's probably worth including a short-circuit condition along the lines of:

```python
if not annotation_ids:
    return []
```

This is because passing empty `IN` clauses to databases doesn't always behave the way you might expect:

```python
>>> print(session.query(models.Flag.annotation_id).filter(models.Flag.annotation_id.in_([])))
SELECT flag.annotation_id AS flag_annotation_id
FROM flag
WHERE flag.annotation_id != flag.annotation_id
```

Uh...

```sql
postgres=# EXPLAIN SELECT flag.annotation_id AS flag_annotation_id FROM flag WHERE flag.annotation_id != flag.annotation_id;
                       QUERY PLAN
---------------------------------------------------------
 Seq Scan on flag  (cost=0.00..24.50 rows=1154 width=16)
   Filter: (annotation_id <> annotation_id)
(2 rows)
```

Uhhhh...!!!

(Yeah, so avoid `.in_(might_be_empty)`).